### PR TITLE
Add ConstrainedLayoutBuilder.updateShouldRebuild()

### PR DESCRIPTION
**Pull request by tvolkert**
This method controls whether the builder needs to be called again again even if the layout constraints are the same.

By default, the builder will always be called when the widget is updated because the logic in the callback might have changed. However, there are cases where subclasses of ConstrainedLayoutBuilder know that certain property updates only affect paint and not build. In these cases, we lack a way of expressing that the builder callback is not needed -- and we end up doing superfluous work.

This PR gives subclasses the ability to know exactly when the callback needs to be called and when it can be skipped.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/layout_builder.dart [hunk 1]
```diff
@@ -47,12 +47,26 @@ abstract class ConstrainedLayoutBuilder<ConstraintType extends Constraints> exte
   /// The builder must not return null.
   final Widget Function(BuildContext context, ConstraintType constraints) builder;
 
+  /// Whether [builder] needs to be called again even if the layout constraints
+  /// are the same.
+  ///
+  /// When this widget is rebuilt, it is most likely that we need to call
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/layout_builder.dart:53]: nit: avoid "we" -> https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple

File: packages/flutter/lib/src/widgets/layout_builder.dart [hunk 2]
```diff
@@ -47,12 +47,26 @@ abstract class ConstrainedLayoutBuilder<ConstraintType extends Constraints> exte
...
+  bool updateShouldRebuild(covariant ConstrainedLayoutBuilder<ConstraintType> oldWidget) => true;
+
   // updateRenderObject is redundant with the logic in the LayoutBuilderElement below.
 }
 
 class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderObjectElement {
   _LayoutBuilderElement(ConstrainedLayoutBuilder<ConstraintType> super.widget);
 
+  @override
+  ConstrainedLayoutBuilder<ConstraintType> get widget => super.widget as ConstrainedLayoutBuilder<ConstraintType>;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/layout_builder.dart:68]: While this used to be the style, we now recommend against providing specialized getters for performance reasons because most callers don't care about the specific type (especially on the rebuild hot path) but still have to pay for the cast. Instead, just cast it at the two or three sides where you need the specific type. See also "Specializing the getters" in our docs: https://master-api.flutter.dev/flutter/widgets/RenderObjectElement-class.html.

File: packages/flutter/lib/src/widgets/layout_builder.dart [hunk 3]
```diff
@@ -47,12 +47,26 @@ abstract class ConstrainedLayoutBuilder<ConstraintType extends Constraints> exte
...
 
+  /// Whether [builder] needs to be called again even if the layout constraints
+  /// are the same.
+  ///
+  /// When this widget is rebuilt, it is most likely that we need to call
+  /// [builder] to build this widget's child. However, subclasses may provide
+  /// ways in which the widget can be updated without needing to rebuild the
+  /// child. Such subclasses can use this method to tell the framework when the
+  /// child widget should be rebuilt.
+  @protected
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/layout_builder.dart:58]: Maybe document explicitly that the _new_ widget is asked if it requires a rebuild and it is given the _old_ widget. Not the other way around.

File: packages/flutter/lib/src/widgets/layout_builder.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/layout_builder.dart:53]: I find the wording "rebuilt" here a little confusing since this callback isn't on the `Element.rebuild` path. Maybe we can reword this to make it clear that this callback is consulted when the widget configuration is changed (i.e. the `Element.update` path)?

Comment (tvolkert) [reply packages/flutter/lib/src/widgets/layout_builder.dart]: Done

File: packages/flutter/lib/src/widgets/layout_builder.dart (hunk 2 above)
Comment (tvolkert) [reply packages/flutter/lib/src/widgets/layout_builder.dart]: Done

File: packages/flutter/lib/src/widgets/layout_builder.dart (hunk 3 above)
Comment (tvolkert) [reply packages/flutter/lib/src/widgets/layout_builder.dart:69]: Done

File: packages/flutter/lib/src/widgets/layout_builder.dart (hunk 1 above)
Comment (tvolkert) [reply packages/flutter/lib/src/widgets/layout_builder.dart]: Done

**Comment by tvolkert** [comment]
Thanks!  PTAL

**Review by goderbauer — APPROVED** [review]
LGTM
